ideafast/middleware-services
# ByteFlies pipeline

This is a draft pull request revolving the ByteFlies pipeline. Due to the ByteFlies having multiple files to download, this PR includes multiple changes that break the other pipelines (for as far VTT was already broken). This is due to a number of required changes (as discussed in #34) in order to get this to work. 

## A rough list of most important changes
See a more detailed discussion about some of these in Issue #34 
- A `Record` now has a hash instead of filename 
- A `Record` stores a `manufacturer_ref` for referencing all records (and their files) to a single recording - per manufacturer. I.e., ByteFlies has ~10 files per recording, where `manufacturer_ref` links to the `id` of that recording for all.
- the `-meta.json` file relates to a recording (depending on manufacturer). A `-preprocessed.json` file can be generated **per file / Record**. This is a clear distinction.
- files (and `-meta.json` files) are now downloaded in _input/PATIENT_ID/DEVICE_ID_ , instead of flattened in _input_. Then, in `task_prepare_data()` these are translated into _PATIENTID-DEVICEID-START-END_ for DMP. If a folder is not processed (i.e. unfinished), new files can easily be added to existing folders 
- `shared. task_prepare_data()` is abstracted to function across device pipeline (`shared_jobs.prepare_data_folders()`). It checks for (sub)folders in the `input` folder, and proceeds to package them for DMP upload if they are ready to be uploaded (i.e. all Records related to the folder have been successfully downloaded and pre-processed).
   -  note the additional methods in `data_transfer/db/main/py` that allow retrieving a list of unfinished 'folders', a list of `Records` coupled to an _input_ folder, and the max and min dates of data across those records.
- A `patient_id` and `device_id` validator is available in the `utils`. Note that this removes any punctuation (e.g. `- dashes`, and cannot be used in - for example - the inventory call (which stores `device_id`s with dashes). 
- For ByteFlies, their API returns a list of recordings, which need to be individually called to get a full list of files to download. This API call is re-run in the download stage to get the latest download URL.
- The API payload of BTF is deep copied (using `__metadata_copy`) in order to create a Record in the next stage. However, only one `-meta.json` file remains for each recording - even though this metadata is deep copied as many times as there are files to be downloaded for each recording.
- `__get_response()` in _lib/byteflies.py_ has a `timer.sleep(1)` to reduce call to their API. 
- A hash is created from information that makes the recording/file unique. For BTF this is the `recording_id`/`signal_id` and if needed /`algorithm_id`. This hash is temporarily stored in the individual result from `lib.get_list()` and removed before writing the `-meta.json`.
 

### To test
- place the `ideafast-byteflies-devices-full.csv` from Teams in the root folder
- update the `.dtransfer.env` with the new/added credentials for Byteflies (see KeePass)
- comment [the inventory endpoints if statement](https://github.com/ideafast/middleware-services/blob/895e57cc650c85fc5f3eee9374abe40fba51119f/consumer/main.py#L14) (and undo indents)
- run `poetry install`
- run `poetry shell`
- run `poetry run consumer` to get the inventory API up an running
- in another terminal window:
- run `poetry shell`
- run `poetry run compose` to get the MongoDB up and running
- run `python data_transfer/main.py` and see metadata being downloaded, files being downloaded, and being moved into the upload folder.

The pipeline only gets [the latest 2 days](https://github.com/ideafast/middleware-services/blob/895e57cc650c85fc5f3eee9374abe40fba51119f/data_transfer/main.py#L79) - which you can adjust. It does however look across study sites at the moment.

You can test if the `prepare_data_folder()` stage ignores 'unfinished' folders by commenting the [last two tasks](https://github.com/ideafast/middleware-services/blob/895e57cc650c85fc5f3eee9374abe40fba51119f/data_transfer/main.py#L87-L90) - and then edit one of the files in the MongoDB and set `is_processed = False`. Then comment the `download` and `preprocess` tasks and uncomment the `prepare_data_folders`.

### Questions / considerations 
The main consideration I am hesitant about it the current approach in `prepare_data_folder()`. Even though it checks that all `Records` associated with a folder are completed, it assumes all files (linked to individual `Records`) are in that folder. 

The alternative would be to get each file associated to a `Record` (i.e. the `manufacturer_ref-meta.json`, `signal or algorithm_id.csv` and possibly a `signal or algorithm_id-preprocessed.json` files) and check if they are present. However, then we would need to track if a `preprocessed.json` is created, and what the files names were once created - which different per device implementation.

**thoughts?**

### Todo (before merging or at some point)
- [x] transform ByteFlies pipeline to run study sites separately
- [ ] introduce a [Inventory endpoint that returns a list of devices per deviceType](https://github.com/ideafast/stack/issues/9#issuecomment-781258338) - this should/could replace the serial .csv lookup (for both Dreem and ByteFlies).
- [ ] clean ByteFlies pipeline, i.e
    - [ ] remove empty download folder in _input_ 
    - [ ] cache ByteFlies API calls (when downloading files from the same recording - should only be available very briefly!)
- [ ] update Dreem pipeline to adapt to these changes
- [ ] update VTT pipeline to adapt these changes
- [x] potentially split validation and formatting functionality of `format_patient_id` for more flexible use.
